### PR TITLE
Adding test for child device conf mgmt

### PR DESCRIPTION
**Pull request by gligorisaev**
Signed-off-by: Gligor Isaev <glgorisaev@gmail.com>

## Proposed changes

Created test for the child device configuration management

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by rina23q — APPROVED** [review]
LGTM

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+${payl_notify}           '{"status": null,  "path": "", "type":"c8y-configuration-plugin", "reason": null}'
+${payl_exec}             '{"status": "executing", "path": "/home/pi/config1", "type": "config1", "reason": null}'
+${payl_succ}             '{"status": "successful", "path": "/home/pi/config1", "type": "config1", "reason": null}'
+
+
+*** Tasks ***
+Create child device name
+    Create Timestamp                                    #Timestamp is used for unique names
+    Define Child Device name                            #Adding CD in front of the timestamp
+Clean the cloud from devices
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:39]:
```suggestion
Clean devices from the cloud
```

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    @{dir_cont}    List Directories In Directory    /etc/tedge/operations/c8y
+    Should Be Empty   ${dir_cont}
+    @{dir_cont}    List Directories In Directory    /etc/tedge/c8y
+    Should Be Empty   ${dir_cont}
+    @{dir_cont}    List Directories In Directory    /var/tedge
+    Should Be Empty   ${dir_cont}
+Delete child related content
+    Execute Command    sudo rm -rf /etc/tedge/operations/c8y/CD*         #if folder exists, child device will be created
+    Execute Command    sudo rm c8y-configuration-plugin.toml
+    Execute Command    sudo rm -rf /etc/tedge/c8y/*                    #if folder exists, child device will be created
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:106]:
```suggestion
    Execute Command    sudo rm -rf /etc/tedge/c8y/CD*                    #if folder exists, child device will be created
```

Otherwise the parent's `c8y-configuration-plugin.toml` will also be deleted.

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    Child Connection
+    ${rc}=    Execute Command    printf ${config} > c8y-configuration-plugin    return_stdout=False    return_rc=True
+    Should Be Equal    ${rc}    ${0}
+    Write    curl -X PUT http://${PARENT_IP}:${HTTP_PORT}/tedge/file-transfer/${CHILD}/c8y-configuration-plugin \\            #Folder will be created /var/tedge/file-transfer
+    Write   --data-binary @- << EOF
+    Write   files = [
+    Write        { path = '/home/pi/config1', type = 'config1' },
+    Write    ]
+    Write  EOF 
+    Execute Command    sudo apt-get install mosquitto-clients -y
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:184]: This installation might be better in `Prerequisite Child Device` step, isn't it?

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    ${auth}=    Create List    ${user}    ${pass}
+    Create Session    API_Testing    https://${url_tedge}    auth=${auth}
+    ${Get_Response}=    GET On Session    API_Testing    /devicecontrol/operations
+    ${json_response}=    Set Variable    ${Get_Response.json()}
+    @{pd_name}=    Get Value From Json    ${json_response}    $..status
+    ${first}    Get From List    ${pd_name}    0
+    ${second}    Get From List    ${pd_name}    1
+    ${third}    Get From List    ${pd_name}    2
+    ${fourth}    Get From List    ${pd_name}    3
+    Should Be Equal    ${first}    SUCCESSFUL
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:230]: This block seems to be repeating multiple times. Can this be extracted into a common routine?

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    GET Child name                                      #Get the Child name from the cloud
+    Validate child Name                                 #This is to check the existence of the bug: https://github.com/thin-edge/thin-edge.io/issues/1569
+Snapshot from device
+    Request snapshot from child device                  #Using the cloud command: "Get snapshot from device"
+    Child device response on snapshot request           #Child device is sending 'executing' and 'successful' MQTT responses
+    No response from child device on snapshot request   #Tests the failing of request after timeout of 10s  
+Child device config update
+    Send configuration to device                        #Using the cloud command: "Send configuration to device"
+    Child device response on update request             #Child device is sending 'executing' and 'successful' MQTT responses
+    No response from child device on config update      #Tests the failing of request after timeout of 10s
```
Comment (albinsuresh) [inline tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:73]: Since both the positive test case and negative test case are covered here, a few more negative test cases can be added, not necessarily in this PR but as follow-up PRs:

1. Genuine failure use-cases where the child device itself reports a failure
2. An operation timing out even after the child device sends an `executing` response
3. False response from the child device on a `config_snapshot` request where the child device sends a successful response without uploading the config file will still result in a FAILED operation in the cloud.
4. Send a bad URL in a `c8y_DownloadConfigFile` request from the cloud and that response should be a failure as thin-edge won't be able to download that file. In this case, a `tedge/{CHILD_ID}/commands/config_snapshot` request won't even be sent to the child device.

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 4 above)
Comment (gligorisaev) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:230]: Difference is the expected result in the Should be equal, it can be of course in separate keyword but than I need to make the variables on suite level, and this will complicate the definitions so I decided to have it like this, and because they are in the keyword session they are reusable. But if there is some relevant concern and not "cosmetic reason" I can reconsider it

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 1 above)
Comment (gligorisaev) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:39]: Done

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 2 above)
Comment (gligorisaev) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:106]: Done

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 3 above)
Comment (gligorisaev) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:184]: In that case the Documentation has to be changed as well, in the documentation this is part of the bootstrapping/startup procedure. I try to keep the tests as close to docs as possible.

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 5 above)
Comment (gligorisaev) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:73]: Not done in this PR because I need more info about how exactly this can be tested, missing boundary values

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 3 above)
Comment (albinsuresh) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:184]: Well... The bootstrapping or even the handling of those requests doesn't necessarily need `mosquitto-clients`. In a realistic use-case, there will be some daemon on the child device that handles these MQTT interactions and that daemon program is free to choose any MQTT library for that purpose. The examples in the documentation just uses `mosquitto-clients` itself just as an example.

The point that I was trying make is that the MQTT library(`mosquitto-clients` in these tests) is not a requirement for the bootstrapping alone, but required for further config operation handlings as well. So, it's more like a prerequisite for the entire `child device agent` software itself.

Here it's not making much difference whether the installation happens in `Prerequisites` phase or here as this bootstrap is the first interaction that uses this library. But, I hope you got my point.

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 5 above)
Comment (albinsuresh) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:73]: Sure. Some of these negative test cases are documented in the test plans [here](https://github.com/thin-edge/thin-edge.io/issues/1480#issuecomment-1291522381) and [here](https://github.com/thin-edge/thin-edge.io/issues/1544#issuecomment-1323691550). If those cases need more detailing, just lemme know.

File: tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot (hunk 3 above)
Comment (gligorisaev) [reply tests/RobotFramework/config_management/child_conf_mgmt_plugin.robot:184]: We just have discussion with @reubenmiller, will add Suite setup and teardown, will include all "preparations" there
